rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/flow/TbRuleChainInputNode.java

```diff
@@ -48,17 +56,81 @@ public class TbRuleChainInputNode implements TbNode {
 
     private TbRuleChainInputNodeConfiguration config;
     private RuleChainId ruleChainId;
+    private boolean forwardMsgToDefaultRuleChain;
 
     @Override
     public void init(TbContext ctx, TbNodeConfiguration configuration) throws TbNodeException {
         this.config = TbNodeUtils.convert(configuration, TbRuleChainInputNodeConfiguration.class);
-        this.ruleChainId = new RuleChainId(UUID.fromString(config.getRuleChainId()));
+        this.forwardMsgToDefaultRuleChain = config.isForwardMsgToDefaultRuleChain();
+        if (forwardMsgToDefaultRuleChain) {
+            this.ruleChainId = getRootRuleChainId(ctx);
+            return;
+        }
+        UUID ruleChainUUID;
+        try {
+            ruleChainUUID = UUID.fromString(config.getRuleChainId());
+        } catch (Exception e) {
+            throw new TbNodeException("Failed to parse rule chain id: " + config.getRuleChainId(), true);
+        }
+        if (ruleChainUUID.equals(ctx.getSelf().getRuleChainId().getId())) {
+            throw new TbNodeException("The rule chain id and the current rule chain id cannot be equals!", true);
+        }
+        this.ruleChainId = new RuleChainId(ruleChainUUID);
         ctx.checkTenantEntity(ruleChainId);
     }
 
     @Override
     public void onMsg(TbContext ctx, TbMsg msg) {
-        ctx.input(msg, ruleChainId);
+        RuleChainId targetRuleChainId = getRuleChainId(ctx, msg);
+        if (targetRuleChainId.equals(ctx.getSelf().getRuleChainId())) {
+            ctx.tellFailure(msg, new RuntimeException(
+                    "Message cannot be forwarded to the default rule chain since current rule chain is the default rule chain!"));
+            return;
+        }
+        ctx.input(msg, targetRuleChainId);
     }
 
+    @Override
+    public TbPair<Boolean, JsonNode> upgrade(int fromVersion, JsonNode oldConfiguration) throws TbNodeException {
+        boolean hasChanges = false;
+        switch (fromVersion) {
+            case 0:
+                if (!oldConfiguration.has("forwardMsgToDefaultRuleChain")) {
+                    hasChanges = true;
+                    ((ObjectNode) oldConfiguration).put("forwardMsgToDefaultRuleChain", false);
+                }
+                break;
+            default:
+                break;
+        }
+        return new TbPair<>(hasChanges, oldConfiguration);
+    }
+
+    private RuleChainId getRuleChainId(TbContext ctx, TbMsg msg) {
+        if (!forwardMsgToDefaultRuleChain) {
+            return ruleChainId;
+        }
```

Comment:
remove this from this method. This should be handled as the first statement in the onMsg method.

```
 if (!forwardMsgToDefaultRuleChain) {
      ctx.input(msg, ruleChainId);
 }
 ```